src/pydvl/influence/torch/functional.py

```diff
@@ -62,70 +97,78 @@ def hvp(
     return output
 
 
-def batch_hvp_gen(
+def create_batch_hvp_function(
     model: torch.nn.Module,
     loss: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
-    data_loader: DataLoader,
     reverse_only: bool = True,
-) -> Generator[Callable[[TorchTensorContainerType], torch.Tensor], None, None]:
+) -> Callable[
+    [Dict[str, torch.Tensor], torch.Tensor, torch.Tensor, torch.Tensor], torch.Tensor
+]:
     r"""
-    Generates a sequence of batch Hessian-vector product (HVP) computations for the provided model, loss function,
-    and data loader. If \(f_i\) is the model's loss on the \(i\)-th batch and \(\theta\) the model parameters,
-    this is the sequence of the callable matrix vector products for the matrices
-
-    \[\nabla_{\theta}\nabla_{\theta}f_i(\theta), \quad i=1,\dots, \text{num_batches} \]
+    Creates a function to compute Hessian-vector product (HVP) for a given model and loss function, where the
+    Hessian information is computed for a provided batch.
 
-    i.e. iterating over the data_loader, yielding partial function calls for calculating HVPs.
+    This function takes a PyTorch model, a loss function, and an optional boolean parameter. It returns a callable
+    that computes the Hessian-vector product for batches of input data and a given vector. The computation can be
+    performed in reverse mode only, based on the `reverse_only` parameter.
 
     Args:
-        model: The PyTorch model for which the HVP is calculated.
-        loss: The loss function used to calculate the gradient and HVP.
-        data_loader: PyTorch DataLoader object containing the dataset for which the HVP is calculated.
-        reverse_only: Whether to use only reverse-mode autodiff
-            (True, default) or both forward- and reverse-mode autodiff (False).
+        model: The PyTorch model for which the Hessian-vector product is to be computed.
+        loss: The loss function. It should take two
+            torch.Tensor objects as input and return a torch.Tensor.
+        reverse_only (bool, optional): If True, the Hessian-vector product is computed in reverse mode only. Defaults
+            to True.
 
-    Yields:
-        Partial functions `H_{batch}(vec)=hvp(model, loss, inputs, targets, vec)` that when called,
-            will compute the Hessian-vector product H(vec) for the given model and loss in a batch-wise manner, where
-            (inputs, targets) coming from one batch.
-    """
+    Returns:
+        Callable[[torch.Tensor, torch.Tensor, torch.Tensor], torch.Tensor]: A function that takes three torch.Tensor
+        objects - input data (x), target data (y), and a vector (vec) - and returns the Hessian-vector product as a
+        torch.Tensor.
```

Comment:
No need to repeat types in the docstrings.
```suggestion
        A function that takes three `torch.Tensor` objects - input data (`x`), target data (`y`), and a vector
        (`vec`), and returns the Hessian-vector product of the loss evaluated on `x`, `y` times `vec`.
```